jep/0000/README.adoc

```diff
@@ -0,0 +1,324 @@
+We will configure Jenkins in a way that physically separates as much as possible the things that must be snapshotted, and things that will not (see <<data_segregation>> for details).
+
+////
+I think .gitignore content must be designed to be able to evolve over time.
+To allow more flexibility, I think the content should be associated between an essentials release/bom to a given .gitignore content.
+////
+
+. Update `.gitignore` content with current Essentials release.
+. `git add --all`
+. `git commit -m '[Upgrade] From BOM x.y.z to a.b.c'`
++
+[NOTE]
+====
+ The commit log should ideally be made understandable for humans.
+ We will use tags to be able to revert/switch between snapshots in a programmatic reliable way.
+ Each tag name should be designed so that it is clear and easy to link it to a given version of Essentials.
+////
+We need to finish up the work on the BOM to be more precise here.
+////
+====
+
+[[data_segregation]]
+===== Segregate configuration from binaries, variable build data, logs...
+
+We will configure Jenkins to use non default values to better separate the more static and critically important things that must be snapshotted, like the configuration for jobs, or for Jenkins itself, from the more fluctuating data like build data, workspaces, exploded content for WAR or plugins, etc.
+
+To keep things simple, we will keep having a single Docker volume, but introduce an additional level to separate things that must be snapshotted, and things that do not need to.
+Incidentally, this will help keep `.gitignore` short.
+
+Basically, instead of the now usual `/var/jenkins_home`, we will introduce two subdirectories under `/var/jenkins` (which we will refer to as `$JENKINS_DIR`):
+
+* `/var/jenkins/home` (=`$JENKINS_HOME`) for to-be-snapshotted part,
+* and `/var/jenkins/var` for the rest.
+
+////
+I'm not a 100% sure about /var/jenkins/var. We could put all those directly under /var/jenkins, but I fear it becomes a bit dirty over time.
+////
+
+Here is what it will look like:
+
+```
+/var/jenkins/
+├── home
+│   ├── jobs
+│   │   └── the_job # configuration file only
+│   ├── nodes
+│   ├── plugins
+│   ├── secrets
+│   ├── updates
+│   ├── userContent
+│   └── users
+└── var
+    ├── logs # JENKINS-50291
+    │   └── tasks
+    ├── plugins # exploded plugins, using --pluginroot switch
+    ├── jobs # JENKINS-50164
+    │   └── the_job
+    │       ├── builds
+    │       └── workspace
+    └── war # using --webroot
+        ├── META-INF
+        ├── WEB-INF
+        ├── ...
+```
+
+===== What to back up
+
+Thanks to the data segregation explained above, we will be snapshotting (almost) everything under `/var/jenkins/home`.
+
+We still need to have a `.gitignore` file for some things that either cannot be moved elsewhere, or that we do not want to store in the Git repository.
+As said above, this will likely be improved as we go.
+
+[source,gitignore,title=.gitignore]
+----
+plugins/
```

Comment:
As I noted in the draft, this _may_ be overly aggressive. Probably you want to limit this to actual plugin binaries (not `*.disabled`) that are in the Essentials BOM:

```
plugins/structs.jpi
# etc.
```

Reply:
For easier access, copying what I wrote there about this.
----
Jesse was saying:
> People may be expecting that a Git revision (plus master.key) corresponds to a fully reconstructible Jenkins instance.

Well, then they will be deceived... It's not a backup system as we already discussed.

To me that this part is going to be out of the scope of the snapshotting system, exactly for the reason you gave: I consider that the whole plugin list is defined explicitly by the BOM, for Essentials plugin.

I would say that for others, well it's also out of scope, and it will be the evergreen client responsibility to store the versions of **all** plugins present there, to be able to restore the previous version if we rollback.

Reply:
In general, the expanded plugin data should go under `/var/cache/jenkins/plugins/` just like `/var/cache/jenkins/war/`.  This way one does not need to worry about catching the expanded plugin data.

Reply:
You'll want to be explicit with you gitignore paths here.  For example, the following gitignore

```
plugins/
```
will ignore all subfolders whose name is plugins.  For example, lets say someone creates a job named `plugins` or a Jenkins folder named `plugins`.  The gitignore as-is will ignore those critical configurations.  gitignore applies to all subdirectories and not just the root.  Instead, use a full path relative to the root of the repository.  Example gitignore,

```
/plugins/
```

Reply:
@samrocketman about the `/var/cache/...` comment can you please post back to the ML to check what/if people have to say about this?

I'm leaning towards changing it this way, but would like to follow the process first to avoid back and forth. Thanks

Reply:
Sure, I don't mind.  It's how I configure my Jenkins installs.

Reply:
Done https://groups.google.com/forum/m/#!topic/jenkinsci-dev/C8fBp3r98bY